pydantic_ai_slim/pydantic_ai/ui/ag_ui/_adapter.py

```diff
@@ -193,3 +202,150 @@ def load_messages(cls, messages: Sequence[Message]) -> list[ModelMessage]:
+            parts_list = list(group)
+
+            if is_assistant:
+                # Combine all parts into a single AssistantMessage
+                content: str | None = None
+                tool_calls: list[ToolCall] = []
+
+                for part in parts_list:
+                    if isinstance(part, TextPart):
+                        content = part.content
+                    elif isinstance(part, ToolCallPart):
+                        tool_calls.append(AGUIAdapter._convert_tool_call(part))
+                    elif isinstance(part, BuiltinToolCallPart):
+                        tool_calls.append(AGUIAdapter._convert_builtin_tool_call(part))
+
+                ag_ui_messages.append(
+                    AssistantMessage(
+                        id=get_next_id(),
+                        content=content,
+                        tool_calls=tool_calls if tool_calls else None,
+                    )
+                )
+            else:
+                # Each non-assistant part becomes its own message
+                for part in parts_list:
+                    if isinstance(part, BuiltinToolReturnPart):
+                        ag_ui_messages.append(
+                            ToolMessage(
+                                id=get_next_id(),
+                                content=AGUIAdapter._serialize_content(part.content),
+                                tool_call_id=AGUIAdapter._make_builtin_tool_call_id(
+                                    part.provider_name, part.tool_call_id
+                                ),
+                            )
+                        )
+
+    @staticmethod
+    def _make_builtin_tool_call_id(provider_name: str | None, tool_call_id: str) -> str:
+        """Create a full builtin tool call ID from provider name and tool call ID."""
+        return f'{BUILTIN_TOOL_CALL_ID_PREFIX}|{provider_name}|{tool_call_id}'
+
+    @staticmethod
+    def _convert_tool_call(part: ToolCallPart) -> ToolCall:
+        """Convert a ToolCallPart to an AG-UI ToolCall."""
+        args_str = part.args if isinstance(part.args, str) else json.dumps(part.args)
+        return ToolCall(
+            id=part.tool_call_id,
+            type='function',
+            function=FunctionCall(
+                name=part.tool_name,
+                arguments=args_str,
+            ),
+        )
+
+    @staticmethod
+    def _convert_builtin_tool_call(part: BuiltinToolCallPart) -> ToolCall:
+        """Convert a BuiltinToolCallPart to an AG-UI ToolCall."""
+        args_str = part.args if isinstance(part.args, str) else json.dumps(part.args)
+        return ToolCall(
+            id=AGUIAdapter._make_builtin_tool_call_id(part.provider_name, part.tool_call_id),
+            type='function',
+            function=FunctionCall(
+                name=part.tool_name,
+                arguments=args_str,
+            ),
+        )
+
+    @staticmethod
+    def _serialize_content(content: Any) -> str:
```

Comment:
Can we use `ToolReturnPart.model_response_str()` and `RetryPromptPart.model_response()`?